Mathlib/Tactic/ToAdditive.lean

```diff
@@ -0,0 +1,467 @@
+      let e' ← etaExpandN (needed_n - es.size) e'
+      return some $ e'
+  trace[to_additive.detail] "expand:\nBefore: {e}\nAfter:  {e₂}"
+  return e₂
+
+def updateWithFun
+  (tgt : Name) (decl : ConstantInfo)
+  : MetaM ConstantInfo := do
+  let mut decl := decl.updateName tgt
+  decl := decl.updateType $ (← applyReplacementFun (← expand decl.type))
+  if let some v := decl.value? then
+    decl := decl.updateValue (← applyReplacementFun (← expand v))
+  return decl
+
+/-- Lean 4 makes declarations which are not internal
+(that is, head string starts with `_`) but which should be transformed.
+eg `proof_1` in `Lean.Meta.mkAuxDefinitionFor` this might be better fixed in core.
+This is a fix for that. -/
+def isInternal' : Name → Bool
+  | n@(Name.str _ s _) => (s.startsWith "proof_") || (Name.isInternal n)
+  | n => Name.isInternal n
+
+/-- transform the declaration `src` and all declarations `pre._proof_i` occurring in `src`
+using the transforms dictionary.
+`replace_all`, `trace`, `ignore` and `reorder` are configuration options.
+`pre` is the declaration that got the `@[to_additive]` attribute and `tgt_pre` is the target of this
+declaration. -/
+partial def transformDeclAux
+  (pre tgt_pre : Name) : Name → CoreM Unit := fun src => do
+  -- if this declaration is not `pre` or an internal declaration, we do nothing.
+  if not (src == pre || isInternal' src) then
+    if (findTranslation? (← getEnv) src).isSome then
+      return ()
+    else
+      throwError "The declaration {pre} depends on the declaration {src} which is in the namespace {pre}, but does not have the `@[to_additive]` attribute. This is not supported. Workaround: move {src} to a different namespace."
+  let env ← getEnv
+  -- we find the additive name of `src`
+  let tgt := src.mapPrefix (fun n => if n == pre then some tgt_pre else none)
+  -- we skip if we already transformed this declaration before
+  if env.contains tgt then
+    return
+  let decl ← getConstInfo src
+  -- we first transform all the declarations of the form `pre._proof_i`
+  for n in decl.type.listNamesWithPrefix pre do
+    transformDeclAux pre tgt_pre n
+  if let some value := decl.value? then
+    for n in value.listNamesWithPrefix pre do
+      transformDeclAux pre tgt_pre n
+  -- we transform `decl` using `f` and the configuration options.
+  let decl : ConstantInfo ← MetaM.run' $ updateWithFun tgt decl
+  -- [todo] above line is simplified, mathlib3 version reads:
+  -- `decl.update_with_fun env (name.map_prefix f) (additive_test f replace_all ignore) relevant reorder tgt`
+  trace[to_additive] "generating\n{decl.name} :=\n  {decl.value!}"
+  try
+    discard <| MetaM.run' <| inferType decl.value!
+  catch
+    | Exception.error stx msg => throwError "@[to_additive] failed.
+      Type mismatch in additive declaration. For help, see the docstring
+      of `to_additive.attr`, section `Troubleshooting`.
+      Failed to add declaration\n{decl.name}:\n{msg}"
+    | e => throwError "unreachable"
```

Comment:
What exactly does this do?

Reply:
It checks whether the decl value typechecks and then throws this slightly different error so that it is clear that the error is coming from to_additive. I tried to copy the exact behaviour that the mathlib3 version uses.